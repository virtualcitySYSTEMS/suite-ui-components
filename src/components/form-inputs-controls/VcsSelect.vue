<template>
  <div
    @mouseover="hover = true"
    @mouseleave="hover = false"
  >
    <VcsTooltip
      :tooltip-position="tooltipPosition"
      :tooltip="errorMessage"
      color="error"
    >
      <template #activator="{ on, attrs }">
        <span v-on="on">
          <v-select
            append-icon="mdi-chevron-down"
            hide-details
            solo
            flat
            :outlined="isOutlined"
            :dense="isDense"
            :height="isDense ? 24 : 32"
            v-bind="{...$attrs, ...attrs}"
            v-on="{...$listeners, ...on}"
            @update:error="setError"
          />
        </span>
      </template>
    </VcsTooltip>
  </div>
</template>
<style lang="scss" scoped>
 .v-text-field {
   padding: 0;
   margin: 0;
 }

 .vcs-select-hover{
   color: var(--v-primary-base) !important;
 }
</style>
<script>

  import VcsTooltip from '../notification/VcsTooltip.vue';
  import validate from '../notification/validation.js';

  /**
   * @description Stylized wrapper around {@link https://vuetifyjs.com/en/api/v-select/ |vuetify select}.
   * Provides two height options depending on "dense" property:
   * - if dense is set true (default), height is 24 px
   * - if dense is set false, height is 32 px
   * Provides VcsTooltip to show error messages
   * @vue-prop {('bottom' | 'left' | 'top' | 'right')}  [tooltipPosition='right'] - Position of the error tooltip.
   * @vue-computed {boolean} isDense - Whether size of select is dense.
   * @vue-computed {boolean} isOutlined - Select is outlined on either hover, focus or error, if not disabled.
   */
  export default {
    name: 'VcsSelect',
    components: { VcsTooltip },
    props: {
      tooltipPosition: {
        type: String,
        default: 'right',
      },
    },
    data() {
      return {
        hover: false,
        errorMessage: '',
      };
    },
    computed: {
      isDense() {
        return this.$attrs.dense !== undefined && this.$attrs.dense !== false;
      },
      isOutlined() {
        return (this.hover || this.errorMessage.length > 0) &&
          !(this.$attrs.disabled || this.$attrs.disabled === '');
      },
    },
    methods: {
      setError() {
        const rules = [...this.$attrs.rules].concat(this.$attrs.errorMessages);
        this.errorMessage = validate(rules, this.$attrs.value).join('\n');
      },
    },
  };
</script>

