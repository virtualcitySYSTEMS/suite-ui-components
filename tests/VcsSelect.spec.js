import { createLocalVue, mount } from '@vue/test-utils';
import { describe, it, expect, beforeAll } from 'vitest';
import Vuetify from 'vuetify';
import VcsSelect from '../src/components/form-inputs-controls/VcsSelect.vue';

describe('VcsSelect', () => {
  let localVue;
  let vuetify;

  beforeAll(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
  });

  it('should mount', () => {
    const vm = mount(VcsSelect, {
      localVue,
      vuetify,
      propsData: {
        items: [{ text: 'foo', value: 'foo' }, { text: 'bar', value: 'bar' }],
        width: 100,
        height: 32,
      },
    });
    expect(vm.html()).toMatchSnapshot();
    vm.destroy();
  });
});
