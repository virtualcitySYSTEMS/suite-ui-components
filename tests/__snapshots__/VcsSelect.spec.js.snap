// Vitest Snapshot v1

exports[`VcsSelect > should mount 1`] = `
"<div data-v-425a9a74=\\"\\" class=\\"v-input vcs-select v-input--hide-details v-input--dense theme--light v-text-field v-text-field--single-line v-text-field--solo v-text-field--solo-flat v-text-field--enclosed v-select\\" style=\\"width: 100px; max-width: 100px;\\">
  <div class=\\"v-input__control\\">
    <div role=\\"button\\" aria-haspopup=\\"listbox\\" aria-expanded=\\"false\\" aria-owns=\\"list-2\\" class=\\"v-input__slot\\" style=\\"height: 32px;\\">
      <div class=\\"v-select__slot\\"><label for=\\"input-2\\" class=\\"v-label theme--light\\" style=\\"left: 0px; position: absolute;\\">Select</label>
        <div class=\\"v-select__selections\\"><input customclasses=\\"\\" width=\\"100\\" id=\\"input-2\\" readonly=\\"readonly\\" type=\\"text\\" aria-readonly=\\"false\\" autocomplete=\\"off\\"></div>
        <div class=\\"v-input__append-inner\\">
          <div class=\\"v-input__icon v-input__icon--append\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-chevron-down theme--light\\"></i></div>
        </div><input type=\\"hidden\\">
      </div>
      <div class=\\"v-menu\\">
        <!---->
      </div>
    </div>
  </div>
</div>"
`;
